glean-core/src/metrics/datetime.rs

```diff
@@ -0,0 +1,118 @@
+use crate::storage::StorageManager;
+use crate::util::get_iso_time_string;
+use crate::CommonMetricData;
+use crate::Glean;
+
+use chrono::{DateTime, FixedOffset, TimeZone};
+
+/// # A datetime metric
+///
+/// Used to record an absolute date and time, such as the time the user first ran
+/// the application.
+#[derive(Debug)]
+pub struct DatetimeMetric {
+    meta: CommonMetricData,
+    time_unit: TimeUnit,
+}
+
+impl MetricType for DatetimeMetric {
+    fn meta(&self) -> &CommonMetricData {
+        &self.meta
+    }
+
+    fn meta_mut(&mut self) -> &mut CommonMetricData {
+        &mut self.meta
+    }
+}
+
+impl DatetimeMetric {
+    pub fn new(meta: CommonMetricData, time_unit: TimeUnit) -> Self {
+        Self { meta, time_unit }
+    }
+
+    /// Public facing API for setting the metric to a date/time which
+    /// includes the timezone offset.
+    ///
+    /// ## Arguments:
+    ///
+    /// * `glean`: the Glean instance this metric belongs to.
+    /// * `year`: the year to set the metric to.
+    /// * `month`: the month to set the metric to (1-12).
+    /// * `day`: the day to set the metric to (1-based).
+    /// * `hour`: the hour to set the metric to.
+    /// * `minute`: the minute to set the metric to.
+    /// * `second`: the second to set the metric to.
+    /// * `nano`: the nanosecond fraction to the last whole second.
+    /// * `offset_seconds`: the timezone difference, in seconds, for the Eastern
+    ///   Hemisphere. Negative seconds mean Western Hemisphere.
+    pub fn set_with_details(
+        &self,
+        glean: &Glean,
+        year: i32,
+        month: u32,
+        day: u32,
+        hour: u32,
+        minute: u32,
+        second: u32,
+        nano: u32,
+        offset_seconds: i32,
+    ) {
+        let timezone_offset = FixedOffset::east_opt(offset_seconds);
+        if timezone_offset.is_none() {
+            log::warn!(
+                "DatetimeMetric::set: invalid timezone offset {}. Not recording.",
+                offset_seconds
+            );
+            return;
+        };
+
+        let datetime_obj = FixedOffset::east(offset_seconds)
+            .ymd_opt(year, month, day)
+            .and_hms_nano_opt(hour, minute, second, nano);
+
+        match datetime_obj.single() {
+            Some(d) => self.set(glean, d),
+            _ => log::warn!("DatetimeMetric::set: invalid input data. Not recording."),
+        }
+    }
+
+    /// Public facing API for setting the metric to a date/time which
+    /// includes the timezone offset.
+    ///
+    /// ## Arguments:
+    ///
+    /// * `glean`: the Glean instance this metric belongs to.
+    /// * `value`: the date/time value, with offset, to set the metric to.
+    pub fn set(&self, glean: &Glean, value: DateTime<FixedOffset>) {
+        if !self.should_record(glean) {
+            return;
+        }
+
+        let value = Metric::Datetime(value, self.time_unit);
+        glean.storage().record(&self.meta, &value)
```

Comment:
I do like how simple the actualy data storage is.

Reply:
Yup :) It's making me feel good about our choice. I fear Desktop, though :P

Reply:
Curious: Dare I ask why you fear desktop?

Reply:
Depending on how we approach there, we might need to abstract away the "lifetime" from the storage, since these concepts there are a bit more ... confused :)